### PR TITLE
Use Python semantics for special binary methods

### LINKED ISSUE #2056: Support special binary methods ("__add__" etc.) as Python does
I'm not sure if this is a bug report or a feature request, but as of cython 0.27.3 `__radd__`, `__rtruediv__` and other right side magic methods do not work:
```python
In [1]: %load_ext Cython

In [2]: %%cython
   ...: cdef class Example:
   ...:   def __radd__(self, other):
   ...:     print('radd', self, other)
   ...:   def __rtruediv__(self, other):
   ...:     print('rtruediv', self, other)
   ...:

In [3]: 3 + Example()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-b54d8dd2451f> in <module>()
----> 1 3 + Example()

TypeError: unsupported operand type(s) for +: 'int' and '_cython_magic_c1fde2d95cc29b87ed98f696ccfef975.Example'

In [4]: 'foo' / Example()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-1e29a35d9b28> in <module>()
----> 1 'foo' / Example()

TypeError: unsupported operand type(s) for /: 'str' and '_cython_magic_c1fde2d95cc29b87ed98f696ccfef975.Example'

In [5]: import Cython

In [6]: Cython.__version__
Out[6]: '0.27.3'
```
I'm using the following workaround, but it looks ugly:
```python
In [7]: %%cython
   ...: cdef class Example:
   ...:   def __add__(other, self):
   ...:     print('add', self, other)
   ...:   def __truediv__(other, self):
   ...:     print('truediv', self, other)
   ...:

In [8]: 3 + Example()
add <_cython_magic_45c6eacd188f8181199b3c76f32cba5d.Example object at 0x109d7a0f0> 3

In [9]: 'foo' / Example()
truediv <_cython_magic_45c6eacd188f8181199b3c76f32cba5d.Example object at 0x109d7a110> foo
```
and doesn't really work if we switch operands:
```python
In [10]: Example() + 3
add 3 <_cython_magic_45c6eacd188f8181199b3c76f32cba5d.Example object at 0x109d7a110>

In [11]: Example() / 'foo'
truediv foo <_cython_magic_45c6eacd188f8181199b3c76f32cba5d.Example object at 0x109468e60>
```

I'm not familiar with cython codebase, but a quick search through the source, issues and PRs doesn't show anything related.

### PR DESCRIPTION
**Pull request by scoder**
Closes #2056.

### CODE REVIEW

File: Cython/Compiler/ModuleNode.py [hunk 1]
```diff
@@ -2030,6 +2033,52 @@ def generate_richcmp_function(self, scope, code):
...
+            code.putln('#define %s %s' % (func_name, slot.left_slot.slot_code(scope)))
+        else:
+            if slot.left_slot.signature == TypeSlots.binaryfunc:
+                extra_arg = extra_arg_decl = ''
+            elif slot.left_slot.signature == TypeSlots.ternaryfunc:
+                extra_arg = ', extra_arg'
+                extra_arg_decl = ', PyObject* extra_arg'
+            else:
+                error(entry.pos, "Unexpected type lost signature: %s" % slot)
+            def has_slot_method(method_name):
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py:2052]:
```suggestion

            def has_slot_method(method_name):
```

File: Cython/Compiler/ModuleNode.py [hunk 2]
```diff
@@ -2030,6 +2033,52 @@ def generate_richcmp_function(self, scope, code):
...
+                entry = scope.lookup(method_name)
+                if entry and entry.is_special and entry.func_cname:
+                    return "%s(%s%s)" % (entry.func_cname, "right, left" if reverse else "left, right", extra_arg)
+                else:
+                    super = 'Py_TYPE(right)->tp_base' if reverse else 'Py_TYPE(left)->tp_base'
+                    return ('(%s->tp_as_number && %s->tp_as_number->%s)'
+                            ' ? %s->tp_as_number->%s(left, right %s)'
+                            ' : (Py_INCREF(Py_NotImplemented), Py_NotImplemented)') % (
+                        super, super, slot.slot_name, super, slot.slot_name, extra_arg)
+            code.putln(
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py:2065]:
```suggestion

            code.putln(
```

File: Cython/Compiler/ModuleNode.py [hunk 3]
```diff
@@ -2030,6 +2033,52 @@ def generate_richcmp_function(self, scope, code):
...
+                TempitaUtilityCode.load_cached(
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py:2066]: This should use `.load_as_string(…)[1]`. The concrete specialised code is unlikely to be used a second time.

File: Cython/Compiler/ModuleNode.py [hunk 4]
```diff
@@ -2030,6 +2033,52 @@ def generate_richcmp_function(self, scope, code):
...
         code.putln("}")
 
+    def generate_binop_function(self, scope, slot, code):
+        func_name = scope.mangle_internal(slot.slot_name)
+        code.putln()
+        preprocessor_guard = slot.preprocessor_guard_code()
+        if preprocessor_guard:
+            code.putln(preprocessor_guard)
+        if scope.directives['c_api_binop_methods']:
+            code.putln('#define %s %s' % (func_name, slot.left_slot.slot_code(scope)))
```
[the remark this answers is not in the document]
Comment (robertwb) [reply Cython/Compiler/ModuleNode.py:2043]: It feels a bit odd, but yeah, the `#define` doesn't technically need the guard.

File: Cython/Compiler/ModuleNode.py (hunk 3 above)
Comment (robertwb) [reply Cython/Compiler/ModuleNode.py:2068]: Good point.

**Comment by scoder** [comment]
Manually merged in e4e1aa37eb4dcc5f5a8eaf4df23b347f513e63b8

**Comment by scoder** [comment]
This change apparently broke existing code: #3688.

**Comment by scoder** [comment]
Reverted in the 0.29.x branch and enabled by default in the master (3.0) branch.

**Pull request closed without merging**